### ISSUE TITLE
Add DirichletBC interface with MeshTags

### ISSUE THREAD

**Issue by jorgensd** [FEniCS/dolfinx#948]
It would very convenient in both Python and C++ to have the following interface
`DirichletBC(FunctionSpace V, Function u, MeshTag mt, index i)` as we had with `MeshFunction`. 
Right now we have to backwards engineer to get the dofs given the facets and an index.

**Comment by francesco-ballarin** [comment]
@michalhabera if I remember correctly one motivation for the lower level DirichletBC was to handle more easily the case of BCs on a component of a mixed function space. I fail to see how having made the switch from MeshFunction to MeshTags would change the underlying motivation in this case, and allow to bring the old interface back.

**Comment by michalhabera** [comment]
Yes, I am also against the change proposed here. The motivation for new DirichletBC interface was to simplify the polymorphic DirichletBC interface and provide full control to the user.
A bit more sense would make to have `locate_dofs_topological(MeshTags mt, value i)` but its doing something very trivial anyway...
@jorgensd It is not now backwards engineering to get dofs, dof indices are what you need to apply dirichlet BC on algebraic level.

**Comment by jorgensd** [comment]
> Yes, I am also against the change proposed here. The motivation for new DirichletBC interface was to simplify the polymorphic DirichletBC interface and provide full control to the user.
> A bit more sense would make to have `locate_dofs_topological(MeshTags mt, value i)` but its doing something very trivial anyway...
> @jorgensd It is not now backwards engineering to get dofs, dof indices are what you need to apply dirichlet BC on algebraic level.

In Python it is possible to do the meshtag to DirichletBC with 2-3 lines of code, Which isn’t terrible. 
In C++, @chrisrichardson and I was working on a demo with ~25 BCs, Which ends up having alot of code to go from a MeshValueCollection to the dof values. 

I agree that this would be solved by having a `locate_dofs_topological(meshtag mt, value i)`would solve the issue

**Comment by garth-wells** [comment]
@jorgensd Can you post some sample code? It would be helpful to start from the issue to see if there are refinements that could be made.

**Comment by jorgensd** [comment]
Sample code: 
```
  // Set up boundary condition on inner surface
  const std::vector<std::int32_t>& facet_array = facet1.values();
  const std::vector<std::int32_t>& facet_indices = facet1.indices();
  int n2 = std::count(facet_array.begin(), facet_array.end(), 2);
  Eigen::Array<std::int32_t, Eigen::Dynamic, 1> facets2(n2);
  int i2 = 0;
  for (std::int32_t i = 0; i < facet_array.size(); ++i)
  {
    if (facet_array[i] == 2)
      facets2[i2++] = facet_indices[i];
    if (i2 > n2)
      throw std::runtime_error("i2 > n2");
  }
  Eigen::Array<std::int32_t, Eigen::Dynamic, 1> dofs2
      = dolfinx::fem::locate_dofs_topological({*V}, mesh->topology().dim() - 1,
                                              facets2);
  auto bc1 = std::make_shared<const fem::DirichletBC>(zero, dofs2);
```

**Comment by michalhabera** [comment]
The "ugly" part now is that of you have a MeshTags object and want to find dofs attached to entities marked with number 42 you need to do, in Python, `dofs = meshtags.indices[meshtags.values == 42]`.

Not too bad in Python, I wonder if there isn't a bit more clever way in c++

**Comment by jorgensd** [comment]
> The "ugly" part now is that of you have a MeshTags object and want to find dofs attached to entities marked with number 42 you need to do, in Python, `dofs = meshtags.indices[meshtags.values == 42]`.
> 
> Not too bad in Python, I wonder if there isn't a bit more clever way in c++

I agree that in Python it is quite easy. For reference you would Need something like this:
```
  bottom_facets = mt.indices[mt.values == 5]
    bottom_dofs = fem.locate_dofs_topological(V, fdim, bottom_facets)
    bc_bottom = fem.DirichletBC(u_bc, bottom_dofs)```

**Comment by garth-wells** [comment]
@michalhabera The Eigen development version has (some) support along the lines of what NumPy can do. I don't know when it will be released. 

The ugly part is indeed extracting data form the MeshTags. Maybe a helper function could be added to `MeshTags`?

**Comment by michalhabera** [comment]
Maybe we can add `MeshTags::indices(T value)` to return only those indices which are marked with value `value`?

**Comment by michalhabera** [comment]
I am very supportive to improve redability and its good @jorgensd is trying things in some real applications. I just do not want to get again into state where DirichletBC is mixing topological information with algebraic and doing operations that could be useful in other context somewhere inside of it.

**Comment by jorgensd** [comment]
> I am very supportive to improve redability and its good @jorgensd is trying things in some real applications. I just do not want to get again into state where DirichletBC is mixing topological information with algebraic and doing operations that could be useful in other context somewhere inside of it.

I think Garth’s suggestion is Good, and can make everything Nice and straightforward.

**Comment by francesco-ballarin** [comment]
Closed by #950
